antsibull/docsite_docs.py

```diff
@@ -0,0 +1,226 @@
+            if not label.startswith(label_prefix):
+                errors.append((
+                    row + 1,
+                    0,
+                    f'Label "{label}" does not start with expected prefix "{label_prefix}"'))
+            else:
+                labels.add(label)
+    return sorted(labels), errors
+
+
+def load_toctree(yaml_section: t.Dict[str, t.Any], section_index: int = 0
+                 ) -> t.Tuple[t.List[str], t.List[str]]:
+    errors: t.List[str] = []
+    toctree: t.List[str] = []
+    if 'toctree' in yaml_section:
+        for toctree_index, toctree_name in enumerate(yaml_section['toctree']):
+            if not isinstance(toctree_name, str):
+                errors.append(
+                    f'Toctree entry #{toctree_index} in section #{section_index}'
+                    f' is not a string')
+                continue
+            toctree.append(toctree_name)
+    return toctree, errors
+
+
+def load_section(yaml_section: t.Dict[str, t.Any], section_index: int = 0
+                 ) -> t.Tuple[t.Optional[Section], t.List[str]]:
+    errors: t.List[str] = []
+    missing = False
+    for required_key in ('title', ):
+        if required_key not in yaml_section:
+            errors.append(
+                f'Section #{section_index} has no "{required_key}" entry')
+            missing = True
+    if missing:
+        return None, errors
+    toctree, toctree_errors = load_toctree(yaml_section, section_index)
+    errors.extend(toctree_errors)
+    if not toctree:
+        errors.append(
+            f'Section #{section_index} has no content')
+        return None, errors
+    return Section(yaml_section['title'], toctree), errors
+
+
+def load_docsite_index(index_path: str, report_not_existing: bool = False
+                       ) -> t.Tuple[t.List[Section], t.List[str]]:
+    '''Load a docsite extra-docs.yml file.
+
+    :arg index_path: Path to extra-docs.yml (does not need to exist).
+    :arg report_not_existing: Whether to report when extra-docs.yml does not exist.
+    :returns: A tuple consisting of a list of sections and a list of error messages.
+    '''
+    sections: t.List[Section] = []
+    errors: t.List[str] = []
+
+    if os.path.isfile(index_path):
+        try:
+            index = load_yaml_file(index_path)
+            if index.get('sections'):
+                for section_index, yaml_section in enumerate(index['sections']):
+                    if not isinstance(yaml_section, dict):
+                        errors.append(f'Section #{section_index} must be a mapping')
+                        continue
+                    section, section_errors = load_section(yaml_section, section_index)
+                    if section is not None:
+                        sections.append(section)
+                    errors.extend(section_errors)
+        except Exception as e:
+            errors.append(str(e))
+    elif report_not_existing:
+        errors.append('extra-docs.yml does not exist')
```

Comment:
I think I'd change this to raise an exception if the rile does not exist and let the calling code decide whether to ignore the error or not.  That simplifies this code in two ways:

* This function does not need to know what the caller intends to do with the results and doesn't need to deviate depending on the callers intentions.
* This function can exit early when index_path is not a file, so we can drop a nesting level.

Reply:
Implemented in 4de665a7a0c51d9ec106adde46ef49f5e7f21e06.